### PR TITLE
Add Dockside

### PR DESCRIPTION
**Pull request by struanb**
[Dockside](https://github.com/newsnowlabs/dockside) is a free and open-source tool for provisioning lightweight access-controlled IDEs, staging environments and sandboxes - aka devtainers - on local machine, self-hosted on-premises on bare metal or VM, or in the cloud.

What sets it apart is:
- its ease of self-hosting and low footprint: launch with either a single `docker run` command, or on Google Cloud with a Deployment Manager script
- the speed with which development environments IDEs can be launched from vanilla or bespoke production images i.e. seconds.
- the fine-grained access control designed for small-to-medium-sized development teams, that allows admins to control what host resources developers and images individual devs can use, and allows devs to control with whom their work is shared
- Dockside is not just a tool for provisioning online IDEs, but also a staging platform: web services running on any development container can be staged either for secure private sharing with colleagues or for publicly sharing with clients or beta testers, all over HTTPS (with self-supplied cert, autogenerated self-signed cert, or a LetsEncrypt certificate automatically provisioned for your own domains).

What also sets Dockside apart is its flexibility to work with pre-existing third-party Docker or system components, such as:
- launching development containers within alternative runtimes like [Gvisor](https://github.com/google/gvisor), [Kata Containers](https://github.com/kata-containers) or [Sysbox](https://github.com/nestybox/sysbox)
- launching containers on x86 platforms that run foreign architecture code, using [qemu-user-static](https://github.com/multiarch/qemu-user-static)
- custom iptables/nftables rules for sandboxing container traffic

### CODE REVIEW

**Review by styfle — COMMENTED** [review]
Is there a cloud/saas version of this? 

"Online IDE" implies it can be accessed online without additional hardware.

**Comment by struanb** [comment]
Hi @styfle thanks very much for considering our pull request.

I am not sure what you mean by "additional hardware" as not all the online IDEs currently listed can in fact be accessed online without additional hardware, or at least not without paid cloud hardware. 

Eclipse Theia, for example, is pure open-source and while it can be self-hosted in the cloud, there is no Saas version of it, so you need additional hardware or cloud instances to host a version you can access online.

For AWS Cloud9 also, you need to launch a paid Amazon cloud instance to run the Cloud9 IDE before you can access it online. It only runs on the paid cloud instance.

Dockside is similar, in that to access it online you can self-host it on a Google, Amazon or Azure cloud instance. We currently include a turnkey script for launching it in Google Cloud, and have a script for Amazon in the works. It can be trivially launched to Azure or other cloud providers too. Like Theia, Dockside is currently a pure open-source project and so there is currently no paid Saas version.

For these reasons, if Theia and AWS Cloud9 qualify for inclusion, then I see no reason why Dockside shouldn't too.

Hope this makes sense!

**Comment by styfle** [comment]
Cloud9 is grandfathered into the list because it was one of the first Cloud IDEs dating back to 2011.

Theia is included in the list as an honorable mention because it is the building block for so many other IDEs.

With the exception of Theia, this is not meant to be a list of IDEs that run in docker. This is a "curated list" of awesome "online IDEs".

The first thing I noticed is that there are 0 issues and 9 PRs for Dockside which tells me this project is still very new.

I try out every IDE before adding to the list (curated) so I'll give it a try once the project matures and we can revisit adding to the awesome list, thanks!

**Comment by struanb** [comment]
Thank you @styfle for your explanation and feedback, we understand that you must curate this list carefully.

The number of issues/PRs on Github do indeed reflects levels of public engagement with Dockside as an open-source project - it having been open-sourced only in October 2021. We have made some progress here, since last December. Dockside now has 178 stars. It has also been featured in console.dev, and blogged about at https://garden.mirahi.io/simplifying-training-set-up-using-container-based-ides-in-dockside/.

I wanted to point out though, for avoidance of doubt, that despite this relatively recent open-source release, the Dockside _software platform_ itself is actually _not_ very new and we do consider it to be quite mature, it having been developed internally and used daily over the past four years by the entire development team at [NewsNow.co.uk](https://www.newsnow.co.uk/)/[NewsNow.com](https://www.newsnow.com/), which is a global top 50 news platform used by millions of people every month. Since last posting here, we have made [5 further public releases](https://github.com/newsnowlabs/dockside/releases) comprising a significant number of fixes and improvements to the platform, including regular upgrades to incorporate newly-released versions of Theia. 

Dockside is also much more than an IDE that runs in Docker. Rather like Gitpod, it is a platform for allowing one's dev team to provision their own online IDEs and staging environments in the cloud. To better explain Dockside, this year we launched the [Dockside.io](https://dockside.io/) homepage, which lists the features and benefits and includes several short screencast videos.

Also, to make Dockside easier to try out, I am also really excited to let you know we have now launched the [Dockside Online IDE Live Demo](https://www.demo.dockside.cloud/). This is now the easiest way to get a flavour of Dockside, as it allows anyone to launch an online demo instance of Dockside in seconds.

I would greatly appreciate you giving the Live Demo a try, and remain hopeful that you will eventually see fit to add Dockside to your list. It is one of only a small number of similar online IDE platforms that can be self-hosted by dev teams (an important USP for organisations that do not allow their code to be hosted by third parties).

Any additional feedback you can provide is also greatly appreciated.

**Comment by styfle** [comment]
Thanks for the description!

I tried out the [Dockside Online IDE Live Demo](https://www.demo.dockside.cloud/) and I see [this](https://user-images.githubusercontent.com/229881/168499585-8101dc04-6ec7-4990-b4b7-09f99c6132cf.png).

So I clicked on "open" next to ide and now I see [this](https://user-images.githubusercontent.com/229881/168499662-cc032d90-0de1-4b54-9b32-99146ce11d4f.png).

Is that expected?

I want to try out all the features to see if it meets the criteria of IDE as opposed to text editor.

**Comment by struanb** [comment]
Hi @styfle. Thanks for trying out the live demo.

Yes - in the first screenshot I can see you've successfully launched a Dockside devtainer _within Dockside_. Doing this is the recursive proof-of-concept, useful for running and developing Dockside-in-Dockside, which is exactly how we develop Dockside. In the second screenshot, you've opened the Theia IDE for your devtainer. The reason it shows code changes is purely incidental (only because the version of Dockside the live demo launches is one that has several uncommited patches specific to the demo environment that are not needed in production).

Some extra things one can try from here:

1. Develop Dockside within the IDE, restart the relevant parts of the stack according to https://github.com/newsnowlabs/dockside/blob/main/docs/developing.md, test your changes by opening the 'inner' Dockside (click 'Open' next to 'dockside') and commit your changes. (You could push if you had SSH keys installed). One can launch multiple Dockside devtainers with different names, for working and testing separate branches simultaneously.
2. Launch a devtainer for a different project, say the dockside.io profile. **This is a good one for testing, as any change you save to `index.html` in the IDE will be visible within some seconds on the staged www service (click 'Open' next to 'www').**

e.g.
image
image
image

**Comment by styfle** [comment]
> Launch a devtainer for a different project, say the dockside.io profile.

Thanks that worked. I was able to change the index.html and it showed up after refreshing the url.

However, the IDE doesn't provide intellisense/autocomplete.

Here's what happens when I type `<input ty` and expected `<input type="" />` to be auto-completed:

image

Compare the same html source code with VS Code intellisense:

image

**Comment by struanb** [comment]
Hi @styfle. The IDE built into Dockside is currently [Eclipse Theia](https://theia-ide.org/).

Theia has some limitations compared to VS Code - it does not come with the same Intellisense out-of-the-box - but it is VS Code compatible, supporting many standard VS Code extensions, which may be pre-installed or installed by a user on-the-fly.

Dockside facilities this. In Dockside, pre-installed extensions may be built into the Docker image for the profile you launch. For example, in the demo Dockside.io profile you've launched, to keep the image small and speedy to load, we currently have preinstalled only these basic VS Code extensions: Docker Language Basics, HTML Language Basics, JSON Language Basics, SCSS Language Basics, and Shell Script Language. They're enough to provide syntax highlighting, but not 

Screenshot showing built-in VS Code extensions in the Dockside.io image:
image

A Dockside administrator can of course build into the image for each profile whatever VS Code extensions they want their dev team to be able to use (any extension at all, which can be downloaded from the Microsoft Extension Marketplace).

A user may then additionally install whatever additional extensions they want (as long as they're available in the https://open-vsx.org/ registry Theia natively supports). For example, enter 'git graph' into the extension search box, and click Install (see 2nd screenshot).

Screenshot showing 'git graph' VS Code extension being installed by the user:
image

N.B. This VS Code extension functionality is largely Theia's, although Dockside facilities it.

So to finally answer your question about intellisense/autocomplete: one has to install a suitable VS Code extension to provide this functionality, which can be done by the user or built by the admin into a profile image. If this is a limitation compared to VS Code, it is a limitation of Theia (and in turn AFAIK all online IDEs based on Theia, such as Gitpod).

There is nothing fundamental that limits Dockside to deploying Theia within devtainers. It could deploy VS Code, and indeed that is a feature we have on the [roadmap](https://github.com/newsnowlabs/dockside/blob/main/docs/roadmap.md).

File: README.md [hunk 1]
```diff
@@ -56,6 +56,7 @@ An ([IDE](https://en.wikipedia.org/wiki/Integrated_development_environment)) is
...
 - [goormIDE](https://ide.goorm.io) <img title="Sign Up Required" width="16" src="https://cdn.jsdelivr.net/npm/simple-icons@1.2.7/icons/shopify.svg" /> - Run almost every programming languages with real-time collaboration, live pair programming, and built-in messenger. 
 - [Coding Rooms](https://codingrooms.com/compiler/) - Edit and run JavaScript, Java, Python, C++ and more in a collaborative environment with intellisense, live chat, audio and video conferencing, real-time editing, and terminal.
+- [Dockside](https://dockside.io/) [<img title="Open Source" width="16" src="https://cdn.jsdelivr.net/npm/simple-icons@1.2.7/icons/github.svg" />](https://github.com/newsnowlabs/dockside) [<img title="Self Hosted" width="16" src="https://cdn.jsdelivr.net/npm/simple-icons@1.2.7/icons/docker.svg" />](https://hub.docker.com/r/newsnowlabs/dockside) - Run parallel access-controlled development containers, from vanilla or production environment images, featuring the Theia IDE (complete with terminals) for developing in any language for multiple architectures, within choice of Docker runtimes; stage running code publicly or securely on own domains over HTTPS.
```
Comment (styfle) [inline README.md:59]:
```suggestion
- [Dockside](https://dockside.io/) [<img title="Open Source" width="16" src="https://cdn.jsdelivr.net/npm/simple-icons@1.2.7/icons/github.svg" />](https://github.com/newsnowlabs/dockside) [<img title="Self Hosted" width="16" src="https://cdn.jsdelivr.net/npm/simple-icons@1.2.7/icons/docker.svg" />](https://hub.docker.com/r/newsnowlabs/dockside) - Self host parallel access-controlled development containers, from vanilla or production environment images, featuring the Theia IDE (complete with terminals) for developing in any language for multiple architectures, within choice of Docker runtimes; stage running code publicly or securely on own domains over HTTPS.
```

Comment (struanb) [reply README.md:59]: Thanks @styfle that's excellent, really happy with this.

Comment (styfle) [reply README.md:59]: If you apply that change then we can go ahead and merge, thanks!

**Review by styfle — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -56,6 +56,7 @@ An ([IDE](https://en.wikipedia.org/wiki/Integrated_development_environment)) is
...
 - [goormIDE](https://ide.goorm.io) <img title="Sign Up Required" width="16" src="https://cdn.jsdelivr.net/npm/simple-icons@1.2.7/icons/shopify.svg" /> - Run almost every programming languages with real-time collaboration, live pair programming, and built-in messenger. 
 - [Coding Rooms](https://codingrooms.com/compiler/) - Edit and run JavaScript, Java, Python, C++ and more in a collaborative environment with intellisense, live chat, audio and video conferencing, real-time editing, and terminal.
+- [Dockside](https://dockside.io/) [<img title="Open Source" width="16" src="https://cdn.jsdelivr.net/npm/simple-icons@1.2.7/icons/github.svg" />](https://github.com/newsnowlabs/dockside) [<img title="Self Hosted" width="16" src="https://cdn.jsdelivr.net/npm/simple-icons@1.2.7/icons/docker.svg" />](https://hub.docker.com/r/newsnowlabs/dockside) - Self host parallel access-controlled development containers complete with the Theia IDE and terminal, for developing in any language and multiple architectures, within choice of Docker runtimes; stage running code publicly or securely on own domains over HTTPS.
```
Comment (styfle) [inline README.md:59]:
```suggestion
- [Dockside](https://dockside.io) [<img title="Open Source" width="16" src="https://cdn.jsdelivr.net/npm/simple-icons@1.2.7/icons/github.svg" />](https://github.com/newsnowlabs/dockside) [<img title="Self Hosted" width="16" src="https://cdn.jsdelivr.net/npm/simple-icons@1.2.7/icons/docker.svg" />](https://hub.docker.com/r/newsnowlabs/dockside) - Self host parallel access-controlled development containers complete with the Theia IDE and terminal, for developing in any language and multiple architectures, within choice of Docker runtimes; stage running code publicly or securely on own domains over HTTPS.
```

**Review by struanb — COMMENTED** [review]
My Github app didn't seem to offer to allow me to commit this latest lint fix with a more suitable commit message. I trust you can fix this when you merge the pull request (otherwise let me know and I will try committing again with a better message).

**Pull request merged**